File: web/apps/api/src/scripts/register-drone.ts
import { randomBytes } from "node:crypto";
import "../config.js"; // loads the monorepo-root .env (sets DATABASE_URL)
import { closePool, getPool } from "@parkdrone/db";
import { hashApiKey } from "../auth.js";

/**
 * Register (or re-key) a drone and print a fresh API key.
 *   tsx src/scripts/register-drone.ts <drone_id> [name]
 * The plaintext key is shown ONCE; only its hash is stored.
 */
async function main(): Promise<void> {
  const droneId = process.argv[2];
  const name = process.argv[3] ?? droneId;
  if (!droneId) {
    console.error("usage: register-drone <drone_id> [name]");
    process.exit(1);
  }
  const key = randomBytes(24).toString("hex");
  await getPool().query(
    `INSERT INTO drone (drone_id, name, api_key_hash)
     VALUES ($1, $2, $3)
     ON CONFLICT (drone_id) DO UPDATE SET name = EXCLUDED.name, api_key_hash = EXCLUDED.api_key_hash`,
    [droneId, name, hashApiKey(key)],
  );
  console.log(`registered drone '${droneId}'`);
  console.log(`API key (store securely, shown once): ${key}`);
}

main()
  .catch((e) => {
    console.error(e);
    process.exitCode = 1;
  })
  .finally(() => closePool());

File: web/apps/vision-worker/parkdrone_vision/register_drone.py
"""Register (or re-key) a drone and print a fresh API key (Python port of
apps/api/src/scripts/register-drone.ts).

    python -m parkdrone_vision.register_drone <drone_id> [name]

The plaintext key is shown ONCE; only its SHA-256 hash is stored.
"""
import secrets
import sys

from .api.auth import hash_api_key
from .db import vision_db as db


def main() -> None:
    if len(sys.argv) < 2:
        print("usage: register_drone <drone_id> [name]", file=sys.stderr)
        sys.exit(1)
    drone_id = sys.argv[1]
    name = sys.argv[2] if len(sys.argv) > 2 else drone_id
    key = secrets.token_hex(24)

    conn = db.connect()
    with conn.cursor() as cur:
        cur.execute(
            """INSERT INTO drone (drone_id, name, api_key_hash)
               VALUES (%s, %s, %s)
               ON CONFLICT (drone_id) DO UPDATE
                 SET name = EXCLUDED.name, api_key_hash = EXCLUDED.api_key_hash""",
            (drone_id, name, hash_api_key(key)),
        )
    conn.commit()
    conn.close()
    print(f"registered drone '{drone_id}'")
    print(f"API key (store securely, shown once): {key}")


if __name__ == "__main__":
    main()
